### PR TITLE
Add fwupd_client_upload_report() for use in gnome-firmware

### PR DESCRIPTION
**Pull request by hughsie**
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

**Comment by superm1** [comment]
From the CI failures It looks like you might be using a symbol not present in older glib versions.

File: libfwupd/fwupd-client-sync.c [hunk 1]
```diff
@@ -2595,6 +2595,70 @@ fwupd_client_upload_bytes(FwupdClient *self,
...
+ * @flags: download flags, e.g. %FWUPD_CLIENT_DOWNLOAD_FLAG_NONE
+ * @cancellable: (nullable): optional #GCancellable
+ * @error: (nullable): optional return location for an error
+ *
+ * Uploads a report to a remote server. The [method@Client.set_user_agent] function
+ * should be called before this method is used.
+ *
+ * Returns: (transfer full): a URI (perhaps an empty string), or %NULL for error
+ *
+ * Since: 1.9.20
```
Comment (superm1) [inline libfwupd/fwupd-client-sync.c:2621]: BTW I'm totally fine if this is for 2.0.0 only and 2.0.0 bumps up min glib deps

Comment (hughsie) [reply libfwupd/fwupd-client-sync.c:2621]: No worries; i can use an older API.

**Comment by hughsie** [comment]
@superm1 in case you missed the big hints, this is all for gnome-firmware to use :)

**Comment by superm1** [comment]
> @superm1 in case you missed the big hints, this is all for gnome-firmware to use :)

Of course :)

File: libfwupd/fwupd-common.c [hunk 2]
```diff
@@ -389,300 +389,6 @@ fwupd_build_user_agent_system(void)
...
- * @error: (nullable): optional return location for an error
- *
- * Gets a salted hash of the /etc/machine-id contents. This can be used to
- * identify a specific machine. It is not possible to recover the original
- * machine-id from the machine-hash.
- *
- * Returns: the SHA256 machine hash, or %NULL if the ID is not present
- *
- * Since: 1.0.4
- **/
```
Comment (superm1) [inline libfwupd/fwupd-common.c:404]: need to update transition docs

**Review by superm1 — DISMISSED** [review]

File: libfwupd/fwupd-common.c (hunk 2 above)
Comment (hughsie) [reply libfwupd/fwupd-common.c:404]: Done.

**Comment by hughsie** [comment]
This actually works now

Screenshot from 2024-05-12 20-47-26

File: libfwupd/fwupd-client.c [hunk 3]
```diff
@@ -344,11 +344,23 @@ fwupd_client_set_host_product(FwupdClient *self, const gchar *host_product)
 	fwupd_client_object_notify(self, "host-product");
 }
 
-static void
+/**
+ * fwupd_client_set_host_machine_id:
+ * @self: a #FwupdClient
+ * @host_machine_id: the host machine ID, usually a SHA-256 hash
+ *
+ * Manually sets the machine ID. This is normally only required for self tests.
```
Comment (superm1) [inline libfwupd/fwupd-client.c:352]: I wonder if it's worth using mock instead of adding new code like this.

File: libfwupd/fwupd-client.c [hunk 4]
```diff
@@ -6372,6 +6548,427 @@ fwupd_client_undo_host_security_attr_finish(FwupdClient *self, GAsyncResult *res
...
+		fwupd_client_build_report_metadata(builder, metadata);
+		json_builder_end_object(builder);
+	}
+
+	/* devices */
+	json_builder_set_member_name(builder, "Devices");
+	json_builder_begin_array(builder);
+	for (guint i = 0; i < devices->len; i++) {
+		FwupdDevice *dev = g_ptr_array_index(devices, i);
+		if (!fwupd_device_has_flag(dev, FWUPD_DEVICE_FLAG_UPDATABLE) &&
```
Comment (superm1) [inline libfwupd/fwupd-client.c:6619]: So I'm wondering if it's worth sending even non-updatable devices. You'll be able to get s very good sampling of extra metadata about the "other" devices in a machine thanks to the CPU plugin.

Comment (hughsie) [reply libfwupd/fwupd-client.c:6619]: My concern there is then the data collection rationale goes from a clear "we want to engage with vendor to increase LVFS participation" to  something a lot less easy to explain.

File: libfwupd/fwupd-client.c (hunk 3 above)
Comment (hughsie) [reply libfwupd/fwupd-client.c]: Actually, I've reverted this -- and just used a conditional so that the machine_id can be unset.

**Pull request merged**